toolchain/check/type_structure.cpp

```diff
@@ -0,0 +1,455 @@
+    // and make sure the cursors are swapped back at the bottom of the loop
+    // iteration, so that they compare with the correct `end()`s.
+    auto unswap =
+        llvm::make_scope_exit([&] { std::swap(lhs_cursor, rhs_cursor); });
+    if (*lhs_cursor == Structural::Symbolic) {
+      std::swap(lhs_cursor, rhs_cursor);
+    } else {
+      unswap.release();
+    }
+
+    // Consume the symbolic on the RHS.
+    ++rhs_cursor;
+
+    // There's a Concrete on the LHS; it matches with the Symbolic and is
+    // consumed.
+    if (*lhs_cursor == Structural::Concrete) {
+      ++lhs_cursor;
+      continue;
+    }
+
+    // The symbolic on the RHS is in the same position as a close paren on the
+    // LHS, which means the structures can not match.
+    //
+    // Example:
+    // - ((c))
+    // - ((c?))
+    if (*lhs_cursor == Structural::ConcreteCloseParen) {
+      return false;
+    }
+
+    // There's an open paren on the LHS; the Symbolic matches with everything on
+    // the LHS up to the matching closing paren.
+    CARBON_CHECK(*lhs_cursor == Structural::ConcreteOpenParen);
+    int depth = 0;
+    do {
+      switch (*lhs_cursor) {
+        case Structural::ConcreteOpenParen:
+          depth += 1;
+          break;
+        case Structural::ConcreteCloseParen:
+          depth -= 1;
+          break;
+        case Structural::Concrete:
+          break;
+        case Structural::Symbolic:
+          break;
+      }
+      ++lhs_cursor;
+    } while (depth > 0);
+  }
+
+  return true;
+}
+
+// A class that builds a `TypeStructure` for an `Impl` that represents its self
+// type.
+class TypeStructureBuilder {
+ public:
+  explicit TypeStructureBuilder(Context& context) : context_(context) {}
+
+  auto Run(SemIR::InstId self_inst_id,
+           SemIR::SpecificInterface interface_constraint) -> TypeStructure {
+    CARBON_CHECK(work_list_.empty());
+
+    position_ = 0;
+    first_symbolic_distance_ = TypeStructure::InfiniteDistance;
+    structure_.clear();
+
+    // The self type comes first in the type structure, so we push it last, as
+    // the queue works from the back.
+    Push(interface_constraint);
+
+    auto maybe_self_type_id = TryGetInstIdAsTypeId(self_inst_id);
+    if (std::holds_alternative<SemIR::TypeId>(maybe_self_type_id)) {
+      auto self_type_id = std::get<SemIR::TypeId>(maybe_self_type_id);
+      CARBON_CHECK(self_type_id != SemIR::TypeId::None);
+      Push(self_type_id);
+    } else {
+      CARBON_CHECK(std::holds_alternative<SymbolicType>(maybe_self_type_id));
+      Push(SymbolicType());
+    }
+    BuildTypeStructure();
+
+    return TypeStructure(std::exchange(structure_, {}),
+                         first_symbolic_distance_);
+  }
+
+ private:
+  auto BuildTypeStructure() -> void {
+    while (!work_list_.empty()) {
+      auto next = work_list_.back();
+      work_list_.pop_back();
+
+      if (std::holds_alternative<CloseType>(next)) {
+        auto close_type = std::get<CloseType>(next);
```

Comment:
This can be done more compactly:
```suggestion
      if (auto* close_type = std::get_if<CloseType>(&next)) {
```

Reply:
Done in a couple places.